src/psyclone/psyir/backend/sympy_writer.py

```diff
@@ -115,86 +132,212 @@ def __init__(self, type_map=None):
+    def __getitem__(self, _):
+        '''This function is only here to trick pylint into thinking that
+        the object returned from ``__new__`` is subscriptable, meaning that
+        code like:
+        ``out = SymPyWriter(exp1, exp2); out[1]`` does not trigger
+        a pylint warning about unsubscriptable-object.
+        '''
+
+    # -------------------------------------------------------------------------
+    def _create_type_map(self, list_of_expressions):
+        '''This function creates a dictionary mapping each Reference in any
+        of the expressions to either a SymPy Function (if the reference
         is an array reference) or a Symbol (if the reference is not an
-        array reference).
+        array reference). It defines a new SymPy function for each array,
+        which has a special write method implemented that automatically
+        converts array indices back by combining each three arguments into
+        one expression (i. e. ``a(1,9,2)`` would become ``a(1:9:2)``).
+
+        A new symbol table is created any time this function is called, so
+        it is important to provide all expressions at once for the symbol
+        table to avoid name clashes in any expression.
 
-        :param list_of_expressions: the list of expressions from which all \
-            references are taken and added to the a symbol table to avoid \
+        :param list_of_expressions: the list of expressions from which all
+            references are taken and added to the a symbol table to avoid
             renaming any symbols (so that only member names will be renamed).
-        :type list_of_expressions: list of \
-            :py:class:`psyclone.psyir.nodes.Node`
-        :returns: the dictionary mapping each reference name to a Sympy \
-            data type (Function of Symbol).
-        :rtype: dictionary of string:Sympy-data-type values
+        :type list_of_expressions: List[:py:class:`psyclone.psyir.nodes.Node`]
 
         '''
-        sympy_type_map = {}
+        # Create a new symbol table, so previous symbol will not affect this
+        # new conversion (i.e. this avoids name clashes with a previous
+        # conversion).
+        self._symbol_table = SymbolTable()
         for expr in list_of_expressions:
             for ref in expr.walk(Reference):
                 name = ref.name
-                if name not in sympy_type_map:
-                    if ref.is_array:
-                        sympy_type_map[name] = Function(name)
-                    else:
-                        sympy_type_map[name] = Symbol(name)
+                if name in self._symbol_table:
+                    continue
+                self._symbol_table.find_or_create(name)
+
+                # Test if an array or an array expression is used:
+                if not ref.is_array:
+                    # A simple scalar:
+                    self._sympy_type_map[name] = Symbol(name)
```

Comment:
Would it be better to use the Symbol created by `find_or_create` here rather than create another instance? (I can imagine this causing confusion in future.)

Reply:
OK.